Sources/TuistLoader/Loaders/CachedManifestLoader.swift

```diff
@@ -0,0 +1,183 @@
+/// time a load was performed.
+///
+public class CachedManifestLoader: ManifestLoading {
+    private let manifestLoader: ManifestLoading
+    private let projectDescriptionHelpersHasher: ProjectDescriptionHelpersHashing
+    private let helpersDirectoryLocator: HelpersDirectoryLocating
+    private let cacheDirectory: AbsolutePath
+    private let fileHandler: FileHandling
+    private let decoder: JSONDecoder = JSONDecoder()
+    private let encoder: JSONEncoder = JSONEncoder()
+    private var helpersCache: [AbsolutePath: String?] = [:]
+
+    public convenience init(manifestLoader: ManifestLoading = ManifestLoader()) {
+        self.init(manifestLoader: manifestLoader,
+                  projectDescriptionHelpersHasher: ProjectDescriptionHelpersHasher(),
+                  helpersDirectoryLocator: HelpersDirectoryLocator(),
+                  cacheDirectory: Environment.shared.cacheDirectory.appending(component: "Manifests"),
+                  fileHandler: FileHandler.shared)
+    }
+
+    init(manifestLoader: ManifestLoading,
+         projectDescriptionHelpersHasher: ProjectDescriptionHelpersHashing,
+         helpersDirectoryLocator: HelpersDirectoryLocating,
+         cacheDirectory: AbsolutePath,
+         fileHandler: FileHandling) {
+        self.manifestLoader = manifestLoader
+        self.projectDescriptionHelpersHasher = projectDescriptionHelpersHasher
+        self.helpersDirectoryLocator = helpersDirectoryLocator
+        self.cacheDirectory = cacheDirectory
+        self.fileHandler = fileHandler
+    }
+
+    public func loadConfig(at path: AbsolutePath) throws -> Config {
+        try manifestLoader.loadConfig(at: path)
+    }
+
+    public func loadProject(at path: AbsolutePath) throws -> Project {
+        try load(manifest: .project, at: path) {
+            try manifestLoader.loadProject(at: path)
+        }
+    }
+
+    public func loadWorkspace(at path: AbsolutePath) throws -> Workspace {
+        try load(manifest: .workspace, at: path) {
+            try manifestLoader.loadWorkspace(at: path)
+        }
+    }
+
+    public func loadSetup(at path: AbsolutePath) throws -> [Upping] {
+        try manifestLoader.loadSetup(at: path)
+    }
+
+    public func loadTemplate(at path: AbsolutePath) throws -> Template {
+        try load(manifest: .template, at: path) {
+            try manifestLoader.loadTemplate(at: path)
+        }
+    }
+
+    public func manifests(at path: AbsolutePath) -> Set<Manifest> {
+        manifestLoader.manifests(at: path)
+    }
+
+    // MARK: - Private
+
+    private func load<T: Codable>(manifest: Manifest, at path: AbsolutePath, loader: () throws -> T) throws -> T {
+        let manifestPath = path.appending(component: manifest.fileName)
+        guard fileHandler.exists(manifestPath) else {
+            throw ManifestLoaderError.manifestNotFound(manifest, path)
+        }
+
+        let manifestHash = try calculateManifestHash(for: manifest, at: manifestPath)
```

Comment:
Should we capture the errors thrown here and read the manifest from disk instead of failing the execution?

Reply:
That makes sense, yes the caching layer should fallback to loading from disk and not impact the loading the process.